src/ast/nodes/AwaitExpression.ts

```diff
@@ -9,4 +11,9 @@ export default class AwaitExpression extends NodeBase {
 	hasEffects(options: ExecutionPathOptions) {
 		return super.hasEffects(options) || !options.ignoreReturnAwaitYield();
 	}
+
+	render(code: MagicString, options: RenderOptions) {
```

Comment:
Thinking about this even more, I think `render` itself is the wrong place to collect this information. Drawbacks are that
* It happens at a very late stage of the build process
* The value will only be set once we get to the first top-level await statement which means previous render operations cannot react to that. This is not a problem now but if we ever implement variant B, we probably want this information to be in place for *all* modules before we start rendering at all.

For the current approach I think this could still be ok but if we want to implement variant B at some point, this should probably move into the initialisation or binding phase. If we do not want to add a flag here, we could check if we are top-level by traversing up the parents until we either get to a (arrow)function or to the Program node.
I now this not the best pattern either but if we ever start inlining functions (in which case this information could be wrong), functions with await expression should probably be left out anyway.

Reply:
The reason for having this in `render` is that if an `await` statement were to be tree-shaken, then we effectively no longer use top-level await.

Reply:
That is, this information is for the purposes of the finaliser.

Reply:
Then maybe this information should actually be gathered when the await expression is included for the first time. I just think it would make sense to have this available reliably before any rendering starts so that if we want to switch to variant B, we have both the whole rendering an finalizing phase available to do the necessary transformations.

Reply:
Sure I've moved it out into "include" with a loop to root. I do think the pattern of contextual visitation state would be useful in Rollup though.